Add a package serialization bit to Module for SIL verifier in Package CMO mode.

elsh opened the pull request:
During Package CMO, SIL cloning happens during which
SILOptions::EnableSerializePackage info is lost.

SILVerifier needs this info to determine whether resilience
can be bypassed for decls serialized in a resiliently
built module when Package CMO optimization enabled.

This PR adds SerializePackageEnabled bit to Module format
and uses that in SILVerifier.

Resolves rdar://126157356

elsh commented:
@swift-ci smoke test

aschwaighofer commented on lib/SIL/Verifier/SILVerifier.cpp:
```diff
@@ -153,12 +153,20 @@ static bool isArchetypeValidInFunction(ArchetypeType *A, const SILFunction *F) {
+  // built); this means package serialization was enabled and direct
+  // access should be allowed.
+  auto serializedInResilientMode =
+      expansion == ResilienceExpansion::Minimal && isSerialized;
```
The resilience expansion of a ‘serialized’ function is always minimal. (See getResilienceExpansion() in SILFunction.h)

by adding this condition (‘isSerialized()‘) we will by pass the check even in regular resilient mode (no package cmo enabled). I don’t think this is desirable.

 Can you restrict this bypass to only when the current module is in the same package?

elsh replied:
Updated to check for that as well as whether module was built resiliently when serialized.

elsh replied:
Had to add SerializePackageEnabled bit to Module; the SILOpts.EnableSerializePackage info gets lost during SIL cloning stage of Package CMO, so it needed to be serialized to the binary module. It's added to this SILVerifier check function along with others we discussed.

elsh commented:
@swift-ci smoke test

elsh committed:
During Package CMO, SIL cloning happens during which
SILOptions::EnableSerializePackage info is lost.

SILVerifier needs this info to determine whether resilience
can be bypassed for decls serialized in a resiliently
built module when Package CMO optimization enabled.

This PR adds SerializePackageEnabled bit to Module format
and uses that in SILVerifier.

Resolves rdar://126157356

elsh committed:
Merge branch 'main' into elsh/pkg-sil-verify

elsh commented:
@swift-ci smoke test

elsh committed:
Merge branch 'main' into elsh/pkg-sil-verify

elsh commented:
@swift-ci smoke test

aschwaighofer reviewed (commented):
LGTM. Thank you
